### ISSUE TITLE
MQTTPersistenceDefault is overwriting old cached files.

### ISSUE THREAD

**Issue by FabianGoettl** [eclipse-paho/paho.mqtt.c#618]
Hi,
I am using MQTTPersistenceDefault to store messages while the client is not connected to the Broker.
This works well. Files are written to the cache folder. 

Problem: If I restart the program, old cache files get overwritten with a new message content. The filenames are c-1.msg, c-2.msg, ... .

Can I prevent that by setting an additonal parameter, which specifies the session (e.g. a session id)? Is this prossible without implementing a custom persistence class?

```
static void connect() {
	MQTTAsync_init_options inits = MQTTAsync_init_options_initializer;
	MQTTAsync_global_init(&inits);

	MQTTAsync mqtt;

	MQTTAsync_createOptions create_options = MQTTAsync_createOptions_initializer;
	create_options.sendWhileDisconnected = true;
	create_options.maxBufferedMessages = MAXINT;
	MQTTAsync_createWithOptions(&mqtt, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_DEFAULT, CACHE_FOLDER, &create_options);

	MQTTAsync_connectOptions connect_options = MQTTAsync_connectOptions_initializer;
	connect_options.cleansession = false;
	connect_options.automaticReconnect = true;
	connect_options.connectTimeout = 10;
	connect_options.keepAliveInterval = 20;

	MQTTAsync_connect(mqtt, &connect_options);
	client = mqtt;
}

static void publish_message(const char *msg) {
	MQTTAsync_responseOptions response_options = MQTTAsync_responseOptions_initializer;
	MQTTAsync_send(client, TOPIC, (int)strlen(msg), msg, QOS, false, &response_options);
}

static void disconnect() {
	MQTTAsync_disconnectOptions disconnect_options = MQTTAsync_disconnectOptions_initializer;
	MQTTAsync_disconnect(client, &disconnect_options);
}
```

**Comment by icraggs** [comment]
The old files shouldn't be overwritten until they aren't used any more - that is, when the messages within them are sent.  If everything is working correctly, then you shouldn't need to do anything else.  (I'm wondering if the message id is restored properly)

Do you know when the data is overwritten?  A protocol trace (see the readme) might help to diagnose what is happening.

There is an issue #588, which affects MQTT V5 only.

**Comment by FabianGoettl** [comment]
I guess your are right. The message id is not restored properly. The client restarts with default msgId=1 and overwrites the old files. However, the messages themselves get restored fine and are published on next connect.

I will double-check the change of msgId/seqno in MQTTAsync_restoreCommands() and MQTTAsync_persistCommand().

**Comment by ruantu** [comment]
This issue is caused by #588, check the MQTAsync.c line 944 @icraggs

**Comment by michael-p** [comment]
Are there any news regarding this issue? I'm also seeing the behavior that persisted messages are not correctly restored/resend on restart, even though they have QoS 1 or 2.

**Comment by icraggs** [comment]
@michael-p Issue #588 was fixed in the last release.  If you are having this problem, some more information about your will be useful/essential.  Perhaps you can get a library trace of the scenario (see the README), or have a sample program?

**Comment by keysight-daryl** [comment]
This might actually be the reconnect scramble issue we are facing currently.  I remember while debugging into this after the reinitialized client it would try overwriting the persisted messages before removing them (presumably happens when it is finally actually delivered to the broker) and somewhere in the process of dealing with this situation is where topics got scrambled going to the broker.  At least on the surface it sounds eerily similar to what I saw happening during the scramble in the debugger.

**Comment by icraggs** [comment]
I'll check, extend or add to the persistence tests to look for this - some basic scenarios are covered already.

**Comment by icraggs** [comment]
@keysight-daryl if you have a test scenario that can be added to the CI tests, that would be ideal.

**Comment by keysight-daryl** [comment]
I don't have any formal test for it since like I said in our packaging we removed all paho library testing entirely and how i debugged into it from our application is too vast to separate out with our custom persistence layers on top of the cpp wrapper etc.  but yeah this sounds almost like a direct hit on the issue we faced during reconnects and the topics being sent to the broker getting scrambled thus forcing continual disconnects and reconnects onward until the persistence folder was removed.  I found it by setting a breakpoint in our custom persistence put() call and seeing that it was trying to write to keys that were still in persistence.  To this day in our custom persist I wasn't sure how we were expected to handle this scenario, do we overwrite the original entry or ignore the request?  All I did was add a warning log and decided to just overwrite the entry.  attached a snippet from our custom put call where you see this warning

![Screen Shot 2020-05-02 at 9 46 30 AM](https://user-images.githubusercontent.com/40178586/80865866-e9295700-8c59-11ea-9d18-8f82d523e3a4.png)

**Comment by icraggs** [comment]
I added some tests as part of #441 and it looks ok so far.

**Comment by icraggs** [comment]
The topic 'scrambling' issue is a result of #877.  If there is an issue here, it is a different one.

**Comment by keysight-daryl** [comment]
I would agree as recently we have even seen the topic scrambling to occur with persistence disabled following a web-socket reconnect, it is just a bit less frequent but equally devastating.  Will be good to see if we still have this 'already exists' persistence issue after upgrading to 1.3.4 or is this only fixed in 1.3.5?  If so we might need to patch this into 1.3.4 somehow.

**Comment by icraggs** [comment]
So my testing indicates that this is working.  If there is an issue, it would be better to describe the problem in terms of the scenario that fails - then that scenario can be added as a test.
